Sources/MongoSwift/BSON/DocumentIterator.swift

```diff
@@ -190,6 +212,35 @@ public class DocumentIterator: IteratorProtocol {
 #endif
     }
 
+    /// Internal helper function for copying elements from some source 'bson_t' to a destination 'bson_t' while
+    /// excluding a non-zero number of keys
+    internal static func copyElements(
+        from srcBSON: OpaquePointer,
+        to dstBSON: OpaquePointer,
+        excluding keys: [String]
+    ) throws {
+        guard !keys.isEmpty else {
+            throw InternalError(message: "No keys to exclude, use 'bson_copy' instead")
+        }
+
+        // use strdup from C standard library to copy each input string
+        var cStrings: [UnsafeMutablePointer<CChar>] = try keys.compactMap {
+            guard let argv = strdup($0) else {
```

Comment:
fyi, we spent a while looking into this - [relevant blog post](https://oleb.net/blog/2016/10/swift-array-of-c-strings/)

see the end of the post for the solution that inspired this... that one has body taking in an array of optional pointers so it conveniently ignores the `strdup` failures

Reply:
see also: https://github.com/apple/swift/blob/ddcd8d324a309844b0e1c39bcd5f97ea957aca41/stdlib/private/SwiftPrivate/SwiftPrivate.swift#L60

Reply:
I've reverted `copyElements` back to the solution in the blog post with an additional step where I unwrap the optionals. It's not the most efficient but let me know what you think!

Reply:
Thanks for the thorough overview about why calling C functions in Swift is not desirable. Using your suggestions and the attached links, I've modified `copyElements` to not call any C functions apart from `bson_copy_to_excluding_noinit_va`.

Reply:
Awesome, looks really good. 

Please note that what I said there is reflective of my opinions on this subject and is not necessarily representative of the Swift community as a whole. I was actually reading a thread on the Swift forums the other day where someone said they like to drop into C for a lot of the heavy lifting and only pop up to Swift at the API level. Similarly, in the world of Rust, there are a lot of discussions going on about how much `unsafe` code should be used and in what circumstances. Some people like to write large swaths of their programs in `unsafe` blocks and others disallow it completely in anything they write. There's a whole spectrum of opinions on this subject across languages, and I just so happen to fall on the safer side.

Also, I apologize for coming in here at the last minute and dropping all this feedback after the PR was really close to merging. I know that can be frustrating, but sometimes it happens during multi-stage reviews. I think that fighting scope creep by filing tickets (as you suggested in the other thread) is a good thing to do to prevent any last minute surprises and to keep things from dragging on.